zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,911 @@
+    ) -> impl Fn(&Zip321Payment, &Zip321Payment) -> Ordering + 'a {
+        move |a: &Zip321Payment, b: &Zip321Payment| {
+            let a_addr = a.recipient_address.render(params);
+            let b_addr = b.recipient_address.render(params);
+
+            a_addr
+                .cmp(&b_addr)
+                .then(a.amount.cmp(&b.amount))
+                .then(a.memo.cmp(&b.memo))
+                .then(a.label.cmp(&b.label))
+                .then(a.message.cmp(&b.message))
+                .then(a.other_params.cmp(&b.other_params))
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Zip321Request {
+    payments: Vec<Zip321Payment>,
+}
+
+impl Zip321Request {
+    pub fn normalize<P: consensus::Parameters>(&mut self, params: &P) {
+        for p in &mut self.payments {
+            p.normalize();
+        }
+
+        self.payments
+            .sort_by(Zip321Payment::compare_normalized(params));
+    }
+
+    pub fn normalize_and_eq<P: consensus::Parameters>(
+        params: &P,
+        a: &mut Zip321Request,
+        b: &mut Zip321Request,
+    ) -> bool {
+        a.normalize(params);
+        b.normalize(params);
+
+        a == b
+    }
+
+    /// Convert this request to a URI string.
+    ///
+    /// Returns None if the payment request is empty.
+    pub fn to_uri<P: consensus::Parameters>(&self, params: &P) -> Option<String> {
+        fn push_params(
+            payment: &Zip321Payment,
+            payment_index: Option<usize>,
+            query_params: &mut Vec<String>,
+        ) {
+            if let Some(amount_p) = render::amount_param(payment.amount, payment_index) {
+                query_params.push(amount_p);
+            }
+
+            if let Some(memo_p) = payment
+                .memo
+                .as_ref()
+                .map(|m| render::memo_param(&m, payment_index))
+            {
+                query_params.push(memo_p);
+            }
+
+            if let Some(label_p) = payment
+                .label
+                .as_ref()
+                .map(|m| render::str_param("label", &m, payment_index))
+            {
+                query_params.push(label_p);
+            }
+
+            if let Some(message_p) = payment
+                .message
+                .as_ref()
+                .map(|m| render::str_param("message", &m, payment_index))
+            {
+                query_params.push(message_p);
+            }
+
+            for (name, value) in &payment.other_params {
+                query_params.push(render::str_param(&name, &value, payment_index))
+            }
+        }
+
+        let mut query_params = vec![];
+        if self.payments.len() == 1 {
+            self.payments.get(0).map(|payment| {
+                push_params(payment, None, &mut query_params);
+                format!(
+                    "zcash:{}?{}",
+                    payment.recipient_address.render(params),
+                    query_params.join("&")
+                )
+            })
```

Comment:
We already have:

> A ZIP-321 URI represents a request for the construction of a transaction having
> one or more *payments*. In the case that only a single payment is being
> requested, the recipient address MAY be included in the ``hier-part`` component
> of the RFC 3986 URI; otherwise, multiple recipient addresses can be specified
> using ``addrparam`` parameters with different indices.

Perhaps we should just change that MAY to SHOULD?